File: cwclock-mobile/src/screens/projects/ProjectFormScreen.js
import React, { useEffect, useMemo, useState } from "react";
import { Alert, StyleSheet } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import Screen from "../../components/Screen";
import FormField from "../../components/FormField";
import SelectField from "../../components/SelectField";
import ColorSwatchPicker from "../../components/ColorSwatchPicker";
import Button from "../../components/Button";
import ErrorBanner from "../../components/ErrorBanner";
import { useI18n } from "../../i18n/I18nContext";
import { apiErrorMessage } from "../../i18n/translate";
import theme from "../../theme";
import { listClientsApi } from "../../redux/clients/clients.actions";
import { createProjectApi, updateProjectApi, deleteProjectApi } from "../../redux/projects/projects.actions";

const DEFAULT_COLOR = "#1cb9f7";

const subdivisionsToText = (subdivisions) => (Array.isArray(subdivisions) ? subdivisions.join(", ") : "");
const textToSubdivisions = (text) =>
  text
    .split(",")
    .map((s) => s.trim())
    .filter(Boolean);

// ProjectFormScreen: create (route.params.project absent) or edit (present).
// Ported from cwclock-ui's EditProjectModal.jsx - color uses a fixed swatch
// row instead of a native color picker (see ColorSwatchPicker.js), and
// cross-client transfer is left to the web app, matching ClientFormScreen's
// equivalent trims.
const ProjectFormScreen = ({ route, navigation }) => {
  const { t, locale } = useI18n();
  const dispatch = useDispatch();
  const { orgId } = useSelector((state) => state.session);
  const { items: clients } = useSelector((state) => state.clients);
  const editingProject = route.params?.project;

  const [name, setName] = useState(editingProject?.name || "");
  const [clientId, setClientId] = useState(editingProject?.clientId || "");
  const [color, setColor] = useState(editingProject?.color || DEFAULT_COLOR);
  const [dailyRate, setDailyRate] = useState(editingProject?.dailyRate != null ? String(editingProject.dailyRate) : "");
  const [subdivisionsText, setSubdivisionsText] = useState(subdivisionsToText(editingProject?.subdivisions));
  const [error, setError] = useState(null);
  const [saving, setSaving] = useState(false);

  useEffect(() => {
    if (orgId) dispatch(listClientsApi(orgId));
  }, [orgId, dispatch]);

  const clientItems = useMemo(() => clients.map((c) => ({ value: c.id, label: c.name })), [clients]);

  const handleSave = async () => {
    if (!name.trim() || !clientId) return;
    setError(null);
    setSaving(true);
    const fields = {
      name: name.trim(),
      color,
      dailyRate: dailyRate === "" ? undefined : Number(dailyRate),
      subdivisions: textToSubdivisions(subdivisionsText),
    };
    try {
      if (editingProject) {
        await dispatch(updateProjectApi(orgId, editingProject.id, { ...fields, clientId }));
      } else {
        await dispatch(createProjectApi(orgId, clientId, fields));
      }
      navigation.goBack();
    } catch (e) {
      setError(apiErrorMessage(e, locale));
    } finally {
      setSaving(false);
    }
  };

  const handleDelete = () => {
    Alert.alert(t("projects.deleteProjectTitle"), t("projects.deleteProjectBody", { name: editingProject.name }), [
      { text: t("common.cancel"), style: "cancel" },
      {
        text: t("common.delete"),
        style: "destructive",
        onPress: async () => {
          await dispatch(deleteProjectApi(orgId, editingProject.id));
          navigation.goBack();
        },
      },
    ]);
  };

  return (
    <Screen>
      <FormField label={t("projects.name")} value={name} onChangeText={setName} />
      <SelectField
        label={t("projects.client")}
        value={clientId}
        onValueChange={setClientId}
        items={clientItems}
        placeholder={t("projects.client")}
      />
      <ColorSwatchPicker label={t("projects.color")} value={color} onChange={setColor} />
      <FormField
        label={t("projects.dailyRate")}
        value={dailyRate}
        onChangeText={setDailyRate}
        keyboardType="decimal-pad"
      />
      <FormField
        label={t("projects.subdivisions")}
        value={subdivisionsText}
        onChangeText={setSubdivisionsText}
        placeholder={t("projects.subdivisionsHint")}
      />
      <ErrorBanner message={error} />
      <Button
        title={t("common.save")}
        onPress={handleSave}
        loading={saving}
        disabled={!name.trim() || !clientId}
        style={editingProject ? styles.saveButton : undefined}
      />
      {editingProject && <Button title={t("common.delete")} variant="danger" onPress={handleDelete} />}
    </Screen>
  );
};

const styles = StyleSheet.create({
  saveButton: {
    marginBottom: theme.spacing(1.5),
  },
});

export default ProjectFormScreen;

File: cwclock-mobile/lib/screens/projects/project_form_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../api/api_client.dart';
import '../../i18n/app_localizations.dart';
import '../../models/project.dart';
import '../../providers/clients_provider.dart';
import '../../providers/locale_provider.dart';
import '../../providers/projects_provider.dart';
import '../../providers/session_provider.dart';
import '../../theme.dart';
import '../../widgets/app_button.dart';
import '../../widgets/app_screen.dart';
import '../../widgets/color_swatch_picker.dart';
import '../../widgets/error_banner.dart';
import '../../widgets/form_field.dart';
import '../../widgets/select_field.dart';
import '../../widgets/app_top_bar.dart';

/// Create (project is null) or edit (present). Color uses a fixed swatch row
/// instead of a native color picker. Ported from
/// src/screens/projects/ProjectFormScreen.js.
class ProjectFormScreen extends ConsumerStatefulWidget {
  final Project? project;

  const ProjectFormScreen({super.key, this.project});

  @override
  ConsumerState<ProjectFormScreen> createState() => _ProjectFormScreenState();
}

class _ProjectFormScreenState extends ConsumerState<ProjectFormScreen> {
  late String _name = widget.project?.name ?? '';
  late String _clientId = widget.project?.clientId ?? '';
  late String _color = widget.project?.color ?? defaultProjectColor;
  late String _dailyRate = widget.project?.dailyRate != null ? widget.project!.dailyRate.toString() : '';
  late String _subdivisionsText = subdivisionsToText(widget.project?.subdivisions ?? const []);
  String? _error;
  bool _saving = false;

  @override
  void initState() {
    super.initState();
    final orgId = ref.read(sessionProvider).orgId;
    if (orgId != null) {
      Future.microtask(() => ref.read(clientsProvider.notifier).listClients(orgId));
    }
  }

  Future<void> _handleSave() async {
    if (_name.trim().isEmpty || _clientId.isEmpty) return;
    final orgId = ref.read(sessionProvider).orgId;
    if (orgId == null) return;
    final locale = ref.read(localeProvider);

    setState(() {
      _error = null;
      _saving = true;
    });
    final fields = {
      'name': _name.trim(),
      'color': _color,
      'dailyRate': _dailyRate.isEmpty ? null : num.tryParse(_dailyRate),
      'subdivisions': textToSubdivisions(_subdivisionsText),
    };
    try {
      if (widget.project != null) {
        await ref
            .read(projectsProvider.notifier)
            .updateProject(orgId, widget.project!.id, {...fields, 'clientId': _clientId});
      } else {
        await ref.read(projectsProvider.notifier).createProject(orgId, _clientId, fields);
      }
      if (mounted) Navigator.of(context).pop();
    } catch (e) {
      setState(() => _error = apiErrorMessage(asApiException(e), locale));
    } finally {
      if (mounted) setState(() => _saving = false);
    }
  }

  void _handleDelete() {
    final locale = ref.read(localeProvider);
    final t = translateWith(locale);
    final orgId = ref.read(sessionProvider).orgId;
    if (orgId == null || widget.project == null) return;

    showDialog(
      context: context,
      builder: (dialogContext) => AlertDialog(
        title: Text(t('projects.deleteProjectTitle')),
        content: Text(t('projects.deleteProjectBody', {'name': widget.project!.name})),
        actions: [
          TextButton(onPressed: () => Navigator.pop(dialogContext), child: Text(t('common.cancel'))),
          TextButton(
            onPressed: () async {
              Navigator.pop(dialogContext);
              await ref.read(projectsProvider.notifier).deleteProject(orgId, widget.project!.id);
              if (!mounted) return;
              Navigator.of(context).pop();
            },
            child: Text(t('common.delete'), style: TextStyle(color: AppColors.of(context).danger)),
          ),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    final locale = ref.watch(localeProvider);
    final t = translateWith(locale);
    final clients = ref.watch(clientsProvider).items;
    final clientItems = [for (final c in clients) SelectItem(c.id, c.name)];

    return Scaffold(
      appBar: AppTopBar(title: widget.project != null ? t('projects.editProject') : t('projects.addProject')),
      body: AppScreen(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            AppFormField(label: t('projects.name'), value: _name, onChanged: (v) => setState(() => _name = v)),
            AppSelectField(
              label: t('projects.client'),
              value: _clientId,
              onChanged: (v) => setState(() => _clientId = v),
              items: clientItems,
              placeholder: t('projects.client'),
            ),
            ColorSwatchPicker(
              label: t('projects.color'),
              value: _color,
              onChanged: (v) => setState(() => _color = v),
            ),
            AppFormField(
              label: t('projects.dailyRate'),
              value: _dailyRate,
              onChanged: (v) => setState(() => _dailyRate = v),
              keyboardType: const TextInputType.numberWithOptions(decimal: true),
            ),
            AppFormField(
              label: t('projects.subdivisions'),
              value: _subdivisionsText,
              onChanged: (v) => setState(() => _subdivisionsText = v),
              placeholder: t('projects.subdivisionsHint'),
            ),
            ErrorBanner(message: _error),
            AppButton(
              title: t('common.save'),
              onPressed: _name.trim().isEmpty || _clientId.isEmpty ? null : _handleSave,
              loading: _saving,
              margin: EdgeInsets.only(bottom: AppSpacing.of(1.5)),
            ),
            if (widget.project != null)
              AppButton(title: t('common.delete'), variant: AppButtonVariant.danger, onPressed: _handleDelete),
          ],
        ),
      ),
    );
  }
}
